Modules/_io/fileio.c

```diff
@@ -707,43 +715,56 @@ static PyObject *
 _io_FileIO_readall_impl(fileio *self)
 /*[clinic end generated code: output=faa0292b213b4022 input=dbdc137f55602834]*/
 {
-    struct _Py_stat_struct status;
     Py_off_t pos, end;
     PyObject *result;
     Py_ssize_t bytes_read = 0;
     Py_ssize_t n;
     size_t bufsize;
-    int fstat_result;
 
-    if (self->fd < 0)
+    if (self->fd < 0) {
         return err_closed();
+    }
 
-    Py_BEGIN_ALLOW_THREADS
-    _Py_BEGIN_SUPPRESS_IPH
+    end = self->size_estimated;
+    if (end <= 0) {
+        /* Use a default size and resize as needed. */
+        bufsize = SMALLCHUNK;
+    }
+    else {
+        /* This is probably a real file, so we try to allocate a
+        buffer one byte larger than the rest of the file.  If the
+        calculation is right then we should get EOF without having
+        to enlarge the buffer. */
+        if (end >= _PY_READ_MAX) {
+            bufsize = _PY_READ_MAX;
+        }
+        else {
+            bufsize = Py_SAFE_DOWNCAST(end, Py_off_t, size_t) + 1;
```

Comment:
I don't think that this cast is safe. You should do something like:

```
bufsize = (end <= SIZE_MAX) ? end : SIZE_MAX;
bufsize++;
```

Reply:
I ran into issues in test_largefile on Windows x86 which caused me to add this. Py_off_t is `long long` on that while size_t is int https://github.com/python/cpython/blob/4f1e1dff8928e626924ee1ff70a77af1a14f0f2a/Modules/_io/_iomodule.h#L95-L106

Reply:
Oop, misread this. The `if end >= _PY_READ_MAX` just before should catch this. (`_PY_READ_MAX` <= SIZE_MAX). 

https://github.com/python/cpython/blob/main/Include/internal/pycore_fileutils.h#L65-L76

Reply:
Sorry, in fact the maximum is PY_SSIZE_T_MAX:

```c
bufsize = (size_t)Py_MIN(end, PY_SSIZE_T_MAX);
if (bufsize < PY_SSIZE_T_MAX) {
    bufsize++;
}
```

Reply:
In this case, replace `bufsize = Py_SAFE_DOWNCAST(end, Py_off_t, size_t) + 1;` with just `bufsize = (size_t)end + 1;`. I just dislike `Py_SAFE_DOWNCAST()` macro, it's not safe, the name is misleading.